controller/util/cf_v2_client.go

```diff
@@ -0,0 +1,172 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+
+	model "github.com/kubernetes-incubator/service-catalog/model/service_broker"
+	scmodel "github.com/kubernetes-incubator/service-catalog/model/service_controller"
+)
+
+const (
+	catalogFormatString         = "%s/v2/catalog"
+	serviceInstanceFormatString = "%s/v2/service_instances/%s"
+	bindingFormatString         = "%s/v2/service_instances/%s/service_bindings/%s"
+)
+
+type cfV2BrokerClient struct {
+	broker *scmodel.ServiceBroker
+}
+
+// Verify that client implements the BrokerClient interface.
+var _ BrokerClient = (*cfV2BrokerClient)(nil)
+
+// CreateCFV2BrokerClient creates an instance of BrokerClient for communicating
+// with brokers which implement the Cloud Foundry Service Broker V2 API.
+func CreateCFV2BrokerClient(b *scmodel.ServiceBroker) BrokerClient {
+	return &cfV2BrokerClient{
+		broker: b,
+	}
+}
+
+func (c *cfV2BrokerClient) GetCatalog() (*model.Catalog, error) {
+	url := fmt.Sprintf(catalogFormatString, c.broker.BrokerURL)
+
+	req, err := http.NewRequest("GET", url, nil)
+	req.SetBasicAuth(c.broker.AuthUsername, c.broker.AuthPassword)
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		log.Printf("Failed to fetch catalog from %s\n%v", url, resp)
+		log.Printf("err: %#v", err)
+		return nil, err
+	}
+
+	var catalog model.Catalog
+	err = ResponseBodyToObject(resp, &catalog)
+	if err != nil {
+		log.Printf("Failed to unmarshal catalog: %#v", err)
+		return nil, err
+	}
+
+	return &catalog, nil
+}
+
+func (c *cfV2BrokerClient) CreateServiceInstance(ID string, req *model.ServiceInstanceRequest) (*model.ServiceInstance, error) {
+	jsonBytes, err := json.Marshal(req)
+	if err != nil {
+		return nil, err
+	}
+
+	url := fmt.Sprintf(serviceInstanceFormatString, c.broker.BrokerURL, ID)
+
+	// TODO: Handle the auth
+	createHTTPReq, err := http.NewRequest("PUT", url, bytes.NewReader(jsonBytes))
+	client := &http.Client{}
+	log.Printf("Doing a request to: %s", url)
+	resp, err := client.Do(createHTTPReq)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	// TODO: Align this with the actual response model.
+	si := model.ServiceInstance{}
+	if err := ResponseBodyToObject(resp, &si); err != nil {
+		return nil, err
+	}
+	return &si, nil
+}
+
+func (c *cfV2BrokerClient) UpdateServiceInstance(ID string, req *model.ServiceInstanceRequest) (*model.ServiceInstance, error) {
+	return nil, fmt.Errorf("Not implemented")
+}
+
+func (c *cfV2BrokerClient) DeleteServiceInstance(ID string) error {
+	url := fmt.Sprintf(serviceInstanceFormatString, c.broker.BrokerURL, ID)
+
+	// TODO: Handle the auth
+	deleteHTTPReq, err := http.NewRequest("DELETE", url, nil)
+	if err != nil {
+		log.Printf("Failed to create new HTTP request: %v", err)
+		return err
+	}
+
+	client := &http.Client{}
```

Comment:
same here RE re-using a client. I'd recommend passing a single client to the `cfV2BrokerClient` struct, storing it therein, and accessing it from all this struct's methods.